Index: bindgen.adb
===================================================================
--- llvm-interface/gnat_src/bindgen.adb	(revision 280157)
+++ llvm-interface/gnat_src/bindgen.adb	(working copy)
@@ -457,7 +457,7 @@
 
       if not Bind_For_Library and not CodePeer_Mode then
          WBI ("      procedure s_stalib_adafinal;");
-         Set_String ("      pragma Import (C, s_stalib_adafinal, ");
+         Set_String ("      pragma Import (Ada, s_stalib_adafinal, ");
          Set_String ("""system__standard_library__adafinal"");");
          Write_Statement_Buffer;
       end if;
